### PR TITLE
NEW: image model + upload to s3

### LINKED ISSUE #186: Support for adding images to questions

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #186
Depends on [this](https://github.com/avantifellows/plio-frontend/pull/331) PR
## Test Plan

- [x] Tested locally
- [x] Tested on staging
- [ ] Tested on production
- [x] If changes in DB, update [DB schema](https://docs.google.com/presentation/d/15c45HbHphh4w3guaFV3BgEA5nMKd-4J35JdUKdrpPzo/edit#slide=id.gc8c9ec0b5e_0_0)

### CODE REVIEW

File: docs/ENV.md [hunk 1]
```diff
@@ -70,6 +70,9 @@ AWS secret access key.
 #### `AWS_REGION`
 Region of the AWS IAM user.
 
+#### `AWS_STORAGE_BUCKET_NAME`
```
Comment (dalmia) [inline docs/ENV.md:73]: have these been made different for staging and prod?

File: docs/ENV.md [hunk 2]
```diff
@@ -70,6 +70,9 @@ AWS secret access key.
...
+AWS bucket where Django storages uploads the files
```
Comment (dalmia) [inline docs/ENV.md:74]: seems like some typo here.

File: plio/models.py [hunk 3]
```diff
@@ -6,6 +6,20 @@
 from plio.config import plio_status_choices, item_type_choices, question_type_choices
 
 
+class Image(SafeDeleteModel):
+    _safedelte_policy = SOFT_DELETE
+
+    image_url = models.ImageField("Image", upload_to="images")
```
Comment (dalmia) [inline plio/models.py:12]:
```suggestion
    url = models.ImageField("Image", upload_to="images")
```

Comment (dalmia) [reply plio/models.py:12]: I think we also need to add alt text - all platforms have it and it's an important element of accessibility.

Comment (dalmia) [reply plio/models.py:12]: we can set some default alt text and might not need to provide option to modify it - but at least something should be there.

File: plio/models.py [hunk 4]
```diff
@@ -6,6 +6,20 @@
...
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        db_table = "image"
+
+    def __str__(self):
+        return f"Image id {self.id}"
```
Comment (dalmia) [inline plio/models.py:20]: consistency with others
```suggestion
        return f"{self.id}: {self.url}"
```

File: plio/serializers.py [hunk 5]
```diff
@@ -1,9 +1,20 @@
...
 from users.models import User
 from users.serializers import UserSerializer
 
 
+class ImageSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Image
+        fields = [
+            "id",
+            "image_url",
```
Comment (dalmia) [inline plio/serializers.py:12]:
```suggestion
            "url",
```

File: plio/settings.py [hunk 6]
```diff
@@ -309,3 +309,9 @@
 EMAIL_HOST_PASSWORD = os.environ.get("EMAIL_HOST_PASSWORD")
 
 SMS_DRIVER = os.environ.get("SMS_DRIVER")
+
+# file storage for uploaded images
+DEFAULT_FILE_STORAGE = "storages.backends.s3boto3.S3Boto3Storage"
+AWS_STORAGE_BUCKET_NAME = os.environ.get("AWS_STORAGE_BUCKET_NAME")
+AWS_QUERYSTRING_AUTH = False
```
Comment (dalmia) [inline plio/settings.py:316]: Screenshot 2021-06-13 at 9 38 54 PM
false is for only public s3 buckets right? We shouldn't have it public right?

File: plio/views.py [hunk 7]
```diff
@@ -346,3 +347,24 @@ def duplicate(self, request, pk):
...
+    destroy: Soft delete a image file entry, along
+             with the uploaded S3 file
+    """
+
+    queryset = Image.objects.all()
+    serializer_class = ImageSerializer
+
+    def perform_destroy(self, instance):
+        instance.image_url.delete()
+        instance.delete()
```
Comment (dalmia) [inline plio/views.py:370]: why is this needed if we didn't do this for VideoViewSet?

File: plio/models.py (hunk 3 above)
Comment (deepansh96) [reply plio/models.py:12]: What is alt text?

File: docs/ENV.md (hunk 2 above)
Comment (deepansh96) [reply docs/ENV.md:74]: Storages?
No. Django storages is the package name which handles the s3 storage

File: docs/ENV.md (hunk 1 above)
Comment (deepansh96) [reply docs/ENV.md:73]: no not yet, will do

File: plio/settings.py (hunk 6 above)
Comment (deepansh96) [reply plio/settings.py:316]: If we don't have a public access enabled for the images, the images will not load on the client, or any browser even. Earlier the bucket was not public, it had no policy for any public user to read the files. And testing during that, I was not able to load the files on the plio

File: plio/views.py (hunk 7 above)
Comment (deepansh96) [reply plio/views.py:370]: Added `instance.image_url.delete()` to delete the image file from s3 as well. Without it, only the entry in the table is soft deleted.

File: docs/ENV.md (hunk 1 above)
Comment (deepansh96) [reply docs/ENV.md:73]: done

File: docs/ENV.md (hunk 2 above)
Comment (dalmia) [reply docs/ENV.md:74]: then better to use `django-storages` or else it's weird to read.

File: plio/models.py (hunk 3 above)
Comment (dalmia) [reply plio/models.py]: https://www.w3schools.com/tags/att_img_alt.asp

File: plio/settings.py (hunk 6 above)
Comment (dalmia) [reply plio/settings.py:316]: bucket level private, file level public

File: plio/views.py (hunk 7 above)
Comment (dalmia) [reply plio/views.py]: create unique name for new image file

Comment (dalmia) [reply plio/views.py]: revert not changing from s3 and update comments on frontend about deleting from s3 too

File: plio/views.py [hunk 8]
```diff
@@ -366,5 +366,4 @@ class ImageViewSet(viewsets.ModelViewSet):
     serializer_class = ImageSerializer
 
     def perform_destroy(self, instance):
-        instance.image_url.delete()
         instance.delete()
```
Comment (dalmia) [inline plio/views.py]: this whole thing is no longer needed right?

File: plio/models.py [hunk 9]
```diff
@@ -2,10 +2,45 @@
 from django.db import models
 import string
 import random
+import os
 from safedelete.models import SafeDeleteModel, SOFT_DELETE
 from plio.config import plio_status_choices, item_type_choices, question_type_choices
 
 
+class Image(SafeDeleteModel):
```
Comment (dalmia) [inline plio/models.py:10]: the only thing missing is tests :P

File: plio/permissions.py [hunk 10]
```diff
@@ -14,7 +15,14 @@ def has_permission(self, request, view):
...
     def has_object_permission(self, request, view, obj):
-        """Object-level permissions for plio. This determines whether the request can access a plio instance or not."""
+        """Object-level permissions for plio/item/question. This determines whether the request can access a plio/item/question instance or not."""
+
+        # if the requested object is an Item or Question, filter the Plio record that it belongs to
+        # and use the permission logic on that Plio instance
+        if isinstance(obj, Question):
+            obj = Plio.objects.filter(id=obj.item.plio.id).first()
+        elif isinstance(obj, Item):
+            obj = Plio.objects.filter(id=obj.plio.id).first()
```
Comment (dalmia) [inline plio/permissions.py:25]: was this added just for sanity or did something broke that led you to add this?
the permission class has only been added to `QuestionViewSet` and not `ItemViewSet`-  any reason?
Any reason for not adding this for ImageViewSet too?

File: plio/tests.py [hunk 11]
```diff
@@ -135,3 +135,106 @@ def setUp(self):
...
+
+class ImageTestCase(PlioTestCase):
+    """Tests the Image CRUD functionality."""
+
+    def setUp(self):
+        # seed some plios
+        super().setUp()
+
+        # seed some items and questions
+        self.plio_1 = Plio.objects.filter(name="Plio 1").first()
```
Comment (dalmia) [inline plio/tests.py:148]: is it sure that this test case will run before PlioTestCase? why depend on it instead of creating a new plio instance here itself?

Comment (dalmia) [inline plio/tests.py:148]: we are not using multiple plios in this test case - so, can only keep `plio`

File: plio/tests.py [hunk 12]
```diff
@@ -135,3 +135,106 @@ def setUp(self):
...
+
+        self.item_1_plio_1 = Item.objects.create(
+            plio=self.plio_1, type="question", time=1
+        )
+        self.item_2_plio_1 = Item.objects.create(
+            plio=self.plio_1, type="question", time=2
+        )
+
+        self.question_1_plio_1 = Question.objects.create(item=self.item_1_plio_1)
+        self.question_2_plio_1 = Question.objects.create(item=self.item_2_plio_1)
```
Comment (dalmia) [inline plio/tests.py:158]: on the basis of the comment above - don't need `plio_1` in the names

File: plio/tests.py [hunk 13]
```diff
@@ -135,3 +135,106 @@ def setUp(self):
...
+        """
+        # upload a test image and retrieve the id
+        with open("plio/static/plio/test_image.jpeg", "rb") as img:
+            response = self.client.post(reverse("images-list"), {"url": img})
+        uploaded_image_id = response.json()["id"]
+
+        # update the question with the newly uploaded image
+        response = self.client.put(
+            reverse("questions-detail", args=[self.question_1_plio_1.id]),
+            {"item": self.item_1_plio_1.id, "image": uploaded_image_id},
```
Comment (dalmia) [inline plio/tests.py:172]: item does not need to passed in put, no? it only updates whatever is being passed - so, item should remain unchanged if not passed.

File: plio/tests.py [hunk 14]
```diff
@@ -135,3 +135,106 @@ def setUp(self):
...
+        Tests whether an authorized user can create/upload an image
+        """
+        with open("plio/static/plio/test_image.jpeg", "rb") as img:
+            response = self.client.post(reverse("images-list"), {"url": img})
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+    def test_guest_cannot_upload_image(self):
+        """
+        Tests whether a guest(unauthroized user) should not be able to create/upload an image
```
Comment (dalmia) [inline plio/tests.py:224]:
```suggestion
        Tests whether a guest(unauthorized user) should not be able to create/upload an image
```

File: plio/tests.py [hunk 15]
```diff
@@ -135,3 +135,106 @@ def setUp(self):
...
+        )
+        # the user should be able to update the question
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_user_cannot_attach_image_to_other_user_question(self):
+        """
+        Tests that a user should NOT be able to link images to the questions that were created by some other user
+        """
+        # create a new user
+        new_user = User.objects.create(mobile="+919988776654")
```
Comment (dalmia) [inline plio/tests.py:182]: Vaibhav used the number `+919876543210` as dummy - best to use the same everywhere?

File: plio/tests.py [hunk 16]
```diff
@@ -135,3 +135,106 @@ def setUp(self):
...
+        new_access_token = AccessToken.objects.create(
+            user=new_user,
+            application=self.application,
+            token=random_token,
+            expires=expires,
+            scope=scopes,
+        )
+        # reset and set the new credentials
+        self.client.credentials()
+        self.client.credentials(HTTP_AUTHORIZATION="Bearer " + new_access_token.token)
```
Comment (dalmia) [inline plio/tests.py:198]: shouldn't the second line alone do it?

File: plio/tests.py (hunk 13 above)
Comment (deepansh96) [reply plio/tests.py:172]: Screenshot 2021-06-15 at 8 53 41 PM

File: plio/models.py [hunk 17]
```diff
@@ -35,9 +35,10 @@ def _generate_unique_uuid(self):
 
     def save(self, *args, **kwargs):
         """Image save method. Before uploading, it creates a unique file name for the image."""
-        self.url.name = "".join(
-            [self._generate_unique_uuid(), os.path.splitext(self.url.name)[1]]
-        )
+        if not self.url.name:
```
Comment (dalmia) [inline plio/models.py:38]:
```suggestion
        # to avoid setting a new name every time the model is saved
        if not self.url.name:
```
